src/Server.php

```diff
@@ -1,95 +1,175 @@
-        $this->DS_TTL = $default_ttl;
-        $this->DS_MAX_LENGTH = $max_packet_len;
-        $this->ds_storage = $ds_storage;
+        // @todo validate configuration validate ip, port, max packet and ttl as positive integers
+
+        if (isset($config['server']) && $server = $config['server']) {
+            $this->ip = isset($server['bind']) ? $server['bind'] : $this->ip;
+            $this->port = isset($server['port']) ? $server['port'] : $this->port;
+            $this->ttl = isset($server['ttl']) ? $server['ttl'] : $this->ttl;
+            $this->maxPacketLength = isset($server['max_packet_length']) ? $server['max_packet_length'] : $this->maxPacketLength;
+        }
+
+        $this->resolver = $resolver;
 
         ini_set('display_errors', true);
         ini_set('error_reporting', E_ALL);
 
-        set_error_handler(array($this, 'ds_error'), E_ALL);
+        set_error_handler([$this, 'errorHandler'], E_ALL);
         set_time_limit(0);
 
         if (!extension_loaded('sockets') || !function_exists('socket_create')) {
-            $this->ds_error(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
+            $this->errorHandler(E_USER_ERROR, 'Socket extension or function not found.', __FILE__, __LINE__);
         }
     }
 
-    public function start()
+    /**
+     * Register event subscriber to class.
+     *
+     * @param EventSubscriberInterface $subscriber
+     */
+    public function registerEventSubscriber(EventSubscriberInterface $subscriber)
     {
-        $ds_socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+        $this->eventSubscribes[] = $subscriber;
+    }
 
-        if (!$ds_socket) {
-            $error = sprintf('Cannot create socket (socket error: %s).', socket_strerror(socket_last_error($ds_socket)));
-            $this->ds_error(E_USER_ERROR, $error, __FILE__, __LINE__);
+    /**
+     * Starts DNS server.
+     */
+    public function run()
+    {
+        $socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+
+        if (!$socket) {
+            $error = sprintf(
+                'Cannot create socket (socket error: %s).',
+                socket_strerror(socket_last_error($socket))
+            );
+            $this->errorHandler(E_USER_ERROR, $error, __FILE__, __LINE__);
         }
 
-        if (!socket_bind($ds_socket, $this->DS_IP, $this->DS_PORT)) {
-            $error = sprintf('Cannot bind socket to %s:%d (socket error: %s).', $this->DS_IP, $this->DS_PORT, socket_strerror(socket_last_error($ds_socket)));
-            $this->ds_error(E_USER_ERROR, $error, __FILE__, __LINE__);
+        if (!socket_bind($socket, $this->ip, $this->port)) {
+            $error = sprintf(
+                'Cannot bind socket to %s:%d (socket error: %s).',
+                $this->ip,
+                $this->port,
+                socket_strerror(socket_last_error($socket))
+            );
+            $this->errorHandler(E_USER_ERROR, $error, __FILE__, __LINE__);
         }
 
```

Comment:
Would be nice to have an info message here that just informs the user that the server has started with the binding IP and port.

Reply:
Agree, although I would put that message in ServerCommand class